syntax = "proto3";

import "google/protobuf/any.proto";

option go_package = "../common";

// Insert data
message protobuf_insert_request {
    string keyspace = 1;
    repeated google.protobuf.Any protobufs = 2;
}

// Selelct data
message protobuf_select_request {
    string keyspace = 1;
    string table = 2;
    string column = 3;
    string constraint = 4;
}

// Update data
message protobuf_update_request {
    string keyspace = 1;
    string table = 2;
    string column = 3;
    string constraint = 4;
    string new_value = 5;
}

// Delete data
message protobuf_delete_request {
    string keyspace = 1;
    string table = 2;
    string column = 3;
    string constraint = 4;
}

// Drop Table
message protobuf_droptable_request {
    string keyspace = 1;
    string table = 2;
}

message protobuf_error_response {
    repeated string errs = 1;
}

message protobuf_select_response {
    string response = 1;
    repeated bytes protobufs = 2;
}

service DBGeneric {
    rpc Insert(protobuf_insert_request) returns (protobuf_error_response);
    rpc Select(protobuf_select_request) returns (protobuf_select_response);
    rpc Update(protobuf_update_request) returns (protobuf_error_response);
    rpc Delete(protobuf_delete_request) returns (protobuf_error_response);
    rpc DropTable(protobuf_droptable_request) returns (protobuf_error_response);
}